File: configs/palette.py
"""
The category palette — one definition, 04_FINDINGS_LEDGER.md C31, build item 43.

C31 measured 0 of 8 categories matching between `ingestion/inference.py`'s
`CATEGORY_COLORS_RGB` (which paints the server-rendered PNG) and `App.jsx`'s
`CAT_COLORS` (which paints the legend beside it). Worst was `dense_vegetation`,
Δ(34, 37, 75) — forest green in the overlay, mint green in the legend.

The deltas are small enough to read as opacity variation, which is what makes
misidentification easy: a user comparing legend to overlay does not conclude
"these disagree", they conclude "that patch is a slightly different shade of the
same thing."

WHAT ACTUALLY FAILED. `inference.py` carried this comment above its palette:

    # Category color palette — must match App.jsx's CAT_COLORS for the
    # 7 ML-resolvable categories exactly, so the frontend legend and the
    # PNG overlay agree visually.

A Python comment asserting a JavaScript constant. That is 01_DIAGNOSIS.md §4 S3
exactly — "a rule enforced only by prose fails at the first edit made by someone
who did not read the prose" — and it failed on every one of the eight.

SO COPYING THE VALUES ACROSS WOULD NOT BE A FIX. It would restore agreement
until the next edit, and leave the same comment making the same unenforceable
promise. This module is the single definition; the values reach the frontend by
being emitted into `result.json` (see `palette_for_result()`), never by being
retyped.

SCOPE. `annotate.py` holds a THIRD copy (`CATEGORY_COLORS`), which also
disagrees with this one — `dense_informal_roofing` is (220, 60, 60) there versus
(224, 60, 60) here. It is a standalone annotation tool, not part of the
pipeline → result.json → frontend path C31 measured, so it is deliberately left
alone rather than changed under an item that did not scope it. It is worth its
own item.
"""

# The seven categories the ML model can resolve. These values are authoritative:
# they are what inference.py has always painted into the PNG overlay, so keeping
# them means the overlay does not change appearance — only the legend, which was
# the half that was wrong.
ML_CATEGORY_COLORS_RGB = {
    "dense_informal_roofing":  (224, 60, 60),
    "sparse_informal_roofing": (240, 140, 80),
    "paved_road":              (120, 120, 180),
    "standing_water":          (40, 100, 200),
    "vegetation_clearing":     (210, 200, 80),
    "active_construction":     (200, 80, 200),
    "dense_vegetation":        (60, 180, 80),
}

# Categories that only ever arrive from OSM vector labels, never from the model
# (see osm_dem.py's OVERRIDABLE set). The frontend renders them, so they belong
# in the emitted palette; without them the frontend would still need a private
# map and the single source of truth would be only three-quarters true.
#
# Values taken from annotate.py, which is where these three have always been
# coloured — chosen over inventing new ones so nothing visibly changes for a
# category that already had an established colour somewhere.
OSM_ONLY_CATEGORY_COLORS_RGB = {
    "unpaved_dirt_road":     (180, 140, 80),
    "open_drainage_channel": (80, 160, 220),
    "open_waste":            (140, 100, 60),
}

# Painted where the CAAT threshold rejected every class. Emitted under the key
# "unknown" so the frontend can colour it like any other category rather than
# special-casing it -- item 43 requires the "unknown" row to travel with the
# rest.
UNKNOWN_COLOR_RGB = (96, 96, 128)
UNKNOWN_INDEX = 255

CATEGORY_COLORS_RGB = {**ML_CATEGORY_COLORS_RGB, **OSM_ONLY_CATEGORY_COLORS_RGB}


def rgb_to_hex(rgb) -> str:
    """(r, g, b) -> '#rrggbb'. The wire format; CSS cannot read a tuple."""
    r, g, b = rgb
    return "#%02x%02x%02x" % (int(r), int(g), int(b))


def palette_for_result() -> dict:
    """
    The palette block emitted into `result.json` under `landcover.palette`.

    This is the mechanism that makes drift structurally impossible rather than
    merely currently-absent: the frontend has no palette of its own to drift
    from, because it reads these values at render time. Changing a colour here
    changes the legend with no frontend edit, which is item 43's acceptance
    criterion.

    `source` is carried so a reader looking at a result.json can find the
    definition without grepping, and `format` so a consumer never has to guess
    whether it is holding tuples or strings.
    """
    colors = {name: rgb_to_hex(rgb) for name, rgb in CATEGORY_COLORS_RGB.items()}
    colors["unknown"] = rgb_to_hex(UNKNOWN_COLOR_RGB)
    return {
        "format": "hex_rgb",
        "source": "configs/palette.py",
        "unknown_index": UNKNOWN_INDEX,
        "ml_categories": sorted(ML_CATEGORY_COLORS_RGB),
        "osm_only_categories": sorted(OSM_ONLY_CATEGORY_COLORS_RGB),
        "colors": colors,
    }

File: geowatch-ui/src/palette.js
/*
 * Category colours — read from the result, never defined here.
 * 04_FINDINGS_LEDGER.md C31, build item 43.
 *
 * This file previously did not exist; App.jsx carried a `CAT_COLORS` literal
 * that was supposed to match `ingestion/inference.py`'s `CATEGORY_COLORS_RGB`,
 * on the authority of a Python comment saying so. 0 of 8 matched. Worst was
 * dense_vegetation, Δ(34, 37, 75) — forest green in the server-rendered
 * overlay, mint green in the legend printed next to it.
 *
 * The deltas were small enough to read as opacity variation, which is what made
 * them dangerous: a user comparing legend to overlay would not conclude "these
 * disagree", they would conclude "that patch is a slightly different shade".
 *
 * THE RULE THIS FILE ENFORCES: there is no palette in the frontend. Colours
 * arrive in `result.landcover.palette`, emitted by configs/palette.py. Drift is
 * not prevented by discipline here, it is impossible, because there is nothing
 * on this side to drift.
 *
 * The one exception is LEGACY_FALLBACK below, and it is deliberately not a
 * palette: see its comment.
 */

/*
 * Runs produced BEFORE item 43 carry no `landcover.palette`. Without something
 * to fall back on, opening an archived result would render every category
 * colourless.
 *
 * These are the OLD frontend values, kept verbatim, precisely BECAUSE they are
 * the wrong ones. An archived run was rendered with these colours when it was
 * produced, so showing it with them now is faithful to what the user saw. They
 * must never be used for a run that carries a palette, and `categoryColor()`
 * only reaches them when `landcover.palette` is absent entirely. A test asserts
 * that.
 *
 * This is not a second source of truth. It is a compatibility shim for data
 * that predates the single source, and it is frozen — no colour is ever added
 * here again.
 */
const LEGACY_FALLBACK = Object.freeze({
  dense_informal_roofing: '#e0625a',
  sparse_informal_roofing: '#e8a35a',
  unpaved_dirt_road: '#b48c50',
  paved_road: '#8888c8',
  open_drainage_channel: '#50a0dc',
  standing_water: '#4a90e2',
  vegetation_clearing: '#d8c85a',
  active_construction: '#c878d0',
  dense_vegetation: '#5ed99b',
  open_waste: '#8c6438',
  unknown: '#716fa0',
})

/** Last resort for a category nobody has a colour for at all. */
export const NO_COLOR = '#888888'

/** The emitted palette block, or null for a run that predates item 43. */
export function paletteOf(result) {
  const p = result?.landcover?.palette
  if (!p || typeof p !== 'object') return null
  if (!p.colors || typeof p.colors !== 'object') return null
  return p
}

/** True when this run carries its own palette (i.e. is not legacy). */
export function hasPalette(result) {
  return paletteOf(result) !== null
}

/**
 * The colour for one category, for this run.
 *
 * Resolution order, and the order matters:
 *   1. the palette emitted with the result — always wins when present;
 *   2. the legacy fallback, ONLY when no palette was emitted at all;
 *   3. NO_COLOR.
 *
 * Step 2 is never reached for a modern run, including for a category the
 * palette happens not to list. That is intentional: falling back per-category
 * would mean a run could render half its legend from the backend and half from
 * here, which is the split-brain state C31 described, just finer-grained.
 */
export function categoryColor(result, category) {
  if (!category) return NO_COLOR
  const palette = paletteOf(result)
  if (palette) return palette.colors[category] || NO_COLOR
  return LEGACY_FALLBACK[category] || NO_COLOR
}

/**
 * Categories to draw in the legend, in emitted order, each with its colour.
 *
 * Reads the palette's own category lists rather than a hardcoded array, so a
 * category added to configs/palette.py appears in the legend with no frontend
 * edit — the same property item 43 requires for colours.
 *
 * `unknown` is excluded: it is a rendering state, not a land-cover type, and
 * the legend has never listed it.
 */
export function legendEntries(result) {
  const palette = paletteOf(result)
  const names = palette
    ? Object.keys(palette.colors)
    : Object.keys(LEGACY_FALLBACK)
  return names
    .filter(name => name !== 'unknown')
    .map(name => ({ category: name, color: categoryColor(result, name) }))
}

/** For tests and diagnostics: is this run rendering from legacy colours? */
export function isLegacyRender(result) {
  return !hasPalette(result)
}
